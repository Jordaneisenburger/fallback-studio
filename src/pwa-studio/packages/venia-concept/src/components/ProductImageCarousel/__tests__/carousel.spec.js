import React from 'react';
import testRenderer from 'react-test-renderer';

import Carousel from '../carousel';

jest.mock('src/classify');

const defaultProps = {
    // This order is specifically set to test sorting/filtering. Do not modify.
    images: [
        {
            file: 'thumbnail.png',
            position: 0,
            disabled: true,
            label: 'disabled-thumbnail'
        },
        {
            file: 'thumbnail2.png',
            position: 2,
            disabled: false,
            label: 'test-thumbnail2'
        },
        {
            file: 'thumbnail1.png',
            position: 1,
            disabled: false,
            label: 'test-thumbnail1'
        },
        // These two images w/o position are required to test the sorting.
        {
            file: 'thumbnail3.png',
            disabled: false,
            label: 'test-thumbnail3'
        },
        {
            file: 'thumbnail4.png',
            disabled: false,
            label: 'test-thumbnail4'
        }
    ]
};

test('renders the Carousel component correctly w/ sorted images', () => {
    const component = testRenderer.create(<Carousel {...defaultProps} />);

    expect(component.toJSON()).toMatchSnapshot();

    const buttons = component.root.findAll(
        el => el.type === 'button' && el.children[0].type === 'img'
    );

    // The first button/image should be thumbnail1 according to position prop.
    expect(buttons[0].children[0].props.alt).toEqual('test-thumbnail1');
    expect(buttons[1].children[0].props.alt).toEqual('test-thumbnail2');
});

test('renders active item as main image', () => {
    const component = testRenderer.create(<Carousel {...defaultProps} />);

    const button = component.root.findByProps({ className: 'rootSelected' });
    const activeImageThumbnailAlt = button.children[0].props.alt;

    const activeImage = component.root.findByProps({
        className: 'currentImage'
    });
    const activeImageAlt = activeImage.props.alt;

    expect(activeImageAlt).toEqual('test-thumbnail1');
    expect(activeImageAlt).toEqual(activeImageThumbnailAlt);
});

test('updates main image when non-active item is clicked', () => {
    const component = testRenderer.create(<Carousel {...defaultProps} />);

    // Click the second image
    component.root
        .findByProps({ alt: 'test-thumbnail2' })
        .parent.props.onClick();

    const button = component.root.findByProps({ className: 'rootSelected' });
    const activeImageThumbnailAlt = button.children[0].props.alt;

    const activeImage = component.root.findByProps({
        className: 'currentImage'
    });
    const activeImageAlt = activeImage.props.alt;

    expect(activeImageAlt).toEqual('test-thumbnail2');
    expect(activeImageAlt).toEqual(activeImageThumbnailAlt);
});

test('renders prior image when left chevron is clicked', () => {
    const component = testRenderer.create(<Carousel {...defaultProps} />);

    // Click the second image to set it as active for this test.
    component.root
        .findByProps({ alt: 'test-thumbnail2' })
        .parent.props.onClick();

    const leftButton = component.root.findByProps({
        className: 'chevron-left'
    });
    leftButton.props.onClick();

    const activeImage = component.root.findByProps({
        className: 'currentImage'
    });
    const activeImageAlt = activeImage.props.alt;

    expect(activeImageAlt).toEqual('test-thumbnail1');
});

test('renders last image when left chevron is clicked and first item is active', () => {
    const component = testRenderer.create(<Carousel {...defaultProps} />);

    const leftButton = component.root.findByProps({
        className: 'chevron-left'
    });
    leftButton.props.onClick();

    const activeImage = component.root.findByProps({
        className: 'currentImage'
    });
    const activeImageAlt = activeImage.props.alt;

    expect(activeImageAlt).toEqual('test-thumbnail4');
});

test('renders next image when right chevron is clicked', () => {
    const component = testRenderer.create(<Carousel {...defaultProps} />);

    const leftButton = component.root.findByProps({
        className: 'chevron-right'
    });
    leftButton.props.onClick();

    const activeImage = component.root.findByProps({
        className: 'currentImage'
    });
    const activeImageAlt = activeImage.props.alt;

    expect(activeImageAlt).toEqual('test-thumbnail2');
});

test('renders first image when right chevron is clicked and last item is active', () => {
    const component = testRenderer.create(<Carousel {...defaultProps} />);

    // Click the last image to set it as active for this test.
    component.root
        .findByProps({ alt: 'test-thumbnail4' })
        .parent.props.onClick();

    const leftButton = component.root.findByProps({
        className: 'chevron-right'
    });
    leftButton.props.onClick();

    const activeImage = component.root.findByProps({
        className: 'currentImage'
    });
    const activeImageAlt = activeImage.props.alt;

    expect(activeImageAlt).toEqual('test-thumbnail1');
});

test('renders a transparent main image if no file name is provided', () => {
    const component = testRenderer.create(<Carousel images={[]} />);
    expect(component.toJSON()).toMatchSnapshot();
});

test('sets main image alt as "image-product" if no label is provided', () => {
    const component = testRenderer.create(<Carousel images={[]} />);

    const activeImage = component.root.findByProps({
        className: 'currentImage'
    });
    const activeImageAlt = activeImage.props.alt;

    expect(activeImageAlt).toEqual('image-product');
});
