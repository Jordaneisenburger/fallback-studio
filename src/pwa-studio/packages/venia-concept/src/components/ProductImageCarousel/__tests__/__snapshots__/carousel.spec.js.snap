// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders a transparent main image if no file name is provided 1`] = `
<div
  className="root"
>
  <div
    className="imageContainer"
  >
    <button
      className="chevron-left"
      onClick={[Function]}
    >
      <span
        className="root"
      >
        <svg
          fill="none"
          height={40}
          stroke="currentColor"
          strokeLinecap="round"
          strokeLinejoin="round"
          strokeWidth="2"
          viewBox="0 0 24 24"
          width={40}
          xmlns="http://www.w3.org/2000/svg"
        >
          <polyline
            points="15 18 9 12 15 6"
          />
        </svg>
      </span>
    </button>
    <img
      alt="image-product"
      className="currentImage"
      src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAQAAAAFCAQAAADIpIVQAAAADklEQVR42mNkgAJGIhgAALQABsHyMOcAAAAASUVORK5CYII="
    />
    <button
      className="chevron-right"
      onClick={[Function]}
    >
      <span
        className="root"
      >
        <svg
          fill="none"
          height={40}
          stroke="currentColor"
          strokeLinecap="round"
          strokeLinejoin="round"
          strokeWidth="2"
          viewBox="0 0 24 24"
          width={40}
          xmlns="http://www.w3.org/2000/svg"
        >
          <polyline
            points="9 18 15 12 9 6"
          />
        </svg>
      </span>
    </button>
  </div>
  <div
    className="root"
  />
</div>
`;

exports[`renders the Carousel component correctly w/ sorted images 1`] = `
<div
  className="root"
>
  <div
    className="imageContainer"
  >
    <button
      className="chevron-left"
      onClick={[Function]}
    >
      <span
        className="root"
      >
        <svg
          fill="none"
          height={40}
          stroke="currentColor"
          strokeLinecap="round"
          strokeLinejoin="round"
          strokeWidth="2"
          viewBox="0 0 24 24"
          width={40}
          xmlns="http://www.w3.org/2000/svg"
        >
          <polyline
            points="15 18 9 12 15 6"
          />
        </svg>
      </span>
    </button>
    <img
      alt="test-thumbnail1"
      className="currentImage"
      src="/img/resize/640?url=%2Fmedia%2Fcatalog%2Fproduct%2Fthumbnail1.png"
    />
    <button
      className="chevron-right"
      onClick={[Function]}
    >
      <span
        className="root"
      >
        <svg
          fill="none"
          height={40}
          stroke="currentColor"
          strokeLinecap="round"
          strokeLinejoin="round"
          strokeWidth="2"
          viewBox="0 0 24 24"
          width={40}
          xmlns="http://www.w3.org/2000/svg"
        >
          <polyline
            points="9 18 15 12 9 6"
          />
        </svg>
      </span>
    </button>
  </div>
  <div
    className="root"
  >
    <button
      className="rootSelected"
      onClick={[Function]}
    >
      <img
        alt="test-thumbnail1"
        className="image"
        src="/img/resize/240?url=%2Fmedia%2Fcatalog%2Fproduct%2Fthumbnail1.png"
      />
    </button>
    <button
      className="root"
      onClick={[Function]}
    >
      <img
        alt="test-thumbnail2"
        className="image"
        src="/img/resize/240?url=%2Fmedia%2Fcatalog%2Fproduct%2Fthumbnail2.png"
      />
    </button>
    <button
      className="root"
      onClick={[Function]}
    >
      <img
        alt="test-thumbnail3"
        className="image"
        src="/img/resize/240?url=%2Fmedia%2Fcatalog%2Fproduct%2Fthumbnail3.png"
      />
    </button>
    <button
      className="root"
      onClick={[Function]}
    >
      <img
        alt="test-thumbnail4"
        className="image"
        src="/img/resize/240?url=%2Fmedia%2Fcatalog%2Fproduct%2Fthumbnail4.png"
      />
    </button>
  </div>
</div>
`;
